src/libraries/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs

```diff
@@ -29,26 +29,119 @@ internal enum CGroupVersion { None, CGroup1, CGroup2 };
         /// <summary>The version of cgroup that's being used. Mutated by tests only.</summary>
         internal static readonly CGroupVersion s_cgroupVersion = FindCGroupVersion();
 
+        /// <summary>Path to the found cgroup memory hierarchy mount path, or null if it couldn't be found.</summary>
+        internal static readonly string? s_cgroupMemoryHierarchyMountPath = FindCGroupMemoryHierarchyMountPath(s_cgroupVersion);
+
         /// <summary>Path to the found cgroup memory limit path, or null if it couldn't be found.</summary>
-        internal static readonly string? s_cgroupMemoryLimitPath = FindCGroupMemoryLimitPath(s_cgroupVersion);
+        internal static readonly string? s_cgroupMemoryPath = FindCGroupMemoryPath(s_cgroupVersion);
 
         /// <summary>Tries to read the memory limit from the cgroup memory location.</summary>
         /// <param name="limit">The read limit, or 0 if it couldn't be read.</param>
         /// <returns>true if the limit was read successfully; otherwise, false.</returns>
         public static bool TryGetMemoryLimit(out ulong limit)
         {
-            string? path = s_cgroupMemoryLimitPath;
+            if (s_cgroupVersion == CGroupVersion.CGroup1)
+            {
+                return TryGetMemoryLimitV1(out limit);
+            }
+            else if (s_cgroupVersion == CGroupVersion.CGroup2)
+            {
+                return TryGetMemoryLimitV2(out limit);
+            }
+
+            limit = 0;
+            return false;
+        }
+
+        /// <summary>Tries to read a field of a specified name from the memory.stat file in the current cgroup (cgroup v1 only).</summary>
+        /// <param name="fieldName">Name of the field to read.</param>
+        /// <param name="val">Value of the field or 0 if the field was not found.</param>
+        /// <returns>true if the field was read successfully; otherwise, false.</returns>
+        internal static bool TryGetMemoryStatField(string fieldName, out ulong val)
+        {
+            string? path = s_cgroupMemoryPath;
+            if (path != null)
+            {
+                try
+                {
+                    // Each field name in the memory.stat is separated by one space from its value
+                    fieldName += ' ';
+                    foreach (string line in File.ReadLines(path + "/memory.stat"))
+                    {
+                        if (line.StartsWith(fieldName))
+                        {
+                            bool foundFieldValue = ulong.TryParse(line.AsSpan(fieldName.Length), out val);
+                            return foundFieldValue;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    Debug.Fail($"Failed to read \"{path}/memory.stat\": {e}");
```

Comment:
Would it be better to let the exception to propagate instead of failing silently?

Reply:
When this exception occurs, it means that the cgroups are somehow badly broken on the device. It is hard to imagine how that would happen unless the kernel is broken. I don't think we want to cause user applications to break just because of that. The c++ code also handles failures to read a value as if it was not present / set.

Reply:
If the kernel is broken, all bets are off.

I think that the most likely cause of the exception thrown here is our own bug that this catch sweeps under the rug.

Reply:
The best practice is to avoid defensive catches like this, and allow the bug to be found and fixed.

Reply:
We would also need to update the Process.MinWorkingSet doc - the list of exceptions it can throw.

Reply:
ok, it can stay as is